terraform/persistent/resources/main.tf

```diff
@@ -0,0 +1,105 @@
+# SPDX-FileCopyrightText: 2022-2024 TII (SSRC) and the Ghaf contributors
+# SPDX-License-Identifier: Apache-2.0
+
+provider "azurerm" {
+  # https://github.com/hashicorp/terraform-provider-azurerm/issues/24804
+  skip_provider_registration = true
+  features {}
+}
+
+terraform {
+  required_providers {
+    azurerm = {
+      source = "hashicorp/azurerm"
+    }
+    secret = {
+      source = "numtide/secret"
+    }
+  }
+}
+
+################################################################################
+
+terraform {
+  # Backend for storing terraform state (see ../../state-storage)
+  backend "azurerm" {
+    # resource_group_name and storage_account_name are set by the callee
+    # from command line in terraform init, see terraform-init.sh
+    container_name = "ghaf-infra-tfstate-container"
+    key            = "ghaf-infra-persistent.tfstate"
+  }
+}
+
+################################################################################
+
+# Variables
+variable "persistent_resource_group" {
+  type        = string
+  description = "Parent resource group name"
+}
+
+locals {
+  # Raise an error if workspace is 'default',
+  # this is a workaround to missing asserts in terraform:
+  assert_workspace_not_default = regex(
+    (terraform.workspace == "default") ?
+  "((Force invalid regex pattern)\n\nERROR: workspace 'default' is not allowed" : "", "")
+
+  # Sanitize workspace name:
+  ws = substr(replace(lower(terraform.workspace), "/[^a-z0-9]/", ""), 0, 16)
```

Comment:
I think `/` is not necessary!

Reply:
it is in fact necessary https://developer.hashicorp.com/terraform/language/functions/replace

Reply:
You mean the two `/` in the regex?
It is required. Please see the documentation: https://developer.hashicorp.com/terraform/language/functions/replace:

> _If substring is wrapped in forward slashes, it is treated as a regular expression, ..._

That's exactly what we want, the string to be treated as a regex.

Reply:
I think you are correct! it is treated as a regex..